fast_dp/fast_dp.py

```diff
@@ -219,28 +208,37 @@ def process(self):
 
         # check input frame limits
 
+        start, end = map(int, self._xds_inp['DATA_RANGE'].split())
+        osc = float(self._xds_inp['OSCILLATION_RANGE'])
+        osc_start = float(self._xds_inp['STARTING_ANGLE'])
+
+        if osc == 0.0:
+            raise RuntimeError('grid scan data')
+
         if not self._first_image is None:
-            if self._metadata['start'] < self._first_image:
-                start = self._metadata['start']
-                self._metadata['start'] = self._first_image
-                self._metadata['phi_start'] += self._metadata['phi_width'] * \
-                                               (self._first_image - start)
+            if start < self._first_image:
+                osc_start += osc * (self._first_image - start)
+                start = self._first_image
+                self._xds_inp['STARTING_ANGLE'] = str(osc_start)
+                self._xds_inp['STARTING_FRAME'] = str(start)
 
         if not self._last_image is None:
-            if self._metadata['end'] > self._last_image:
-                self._metadata['end'] = self._last_image
+            if end > self._last_image:
+                end = self._last_image
```

Comment:
```python
if self._last_image:
  end = max(end, self._last_image)
```

Reply:
Actually that should be min() but the idea is sound